elasticapm/metrics/sets/gc.py

```diff
@@ -0,0 +1,74 @@
+#  BSD 3-Clause License
+#
+#  Copyright (c) 2019, Elasticsearch BV
+#  All rights reserved.
+#
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions are met:
+#
+#  * Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+#  * Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+#  * Neither the name of the copyright holder nor the names of its
+#    contributors may be used to endorse or promote products derived from
+#    this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+#  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+#  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+#  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+#  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+#  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+#  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+from __future__ import absolute_import
+
+import gc
+import timeit
+
+from elasticapm.metrics.base_metrics import MetricsSet
+from elasticapm.utils import compat
+
+
+class GarbageMetrics(MetricsSet):
+    START = "start"
+    STOP = "stop"
+
+    def __init__(self, registry):
+        self.previous = self.get_gc_stats()
+        self.gc_start = None
+        if hasattr(gc, "callbacks"):  # only in Python 3.3+
+            gc.callbacks.append(self.gc_callback)
+        super(GarbageMetrics, self).__init__(registry)
+
+    def collect(self):
+        new = self.get_gc_stats()
+        prev, self.previous = self.previous, new
+        delta = {k: new[k] - prev[k] for k in new.keys()}
+        for k, v in compat.iteritems(delta):
+            self.counter(k).inc(v)
+        return super(GarbageMetrics, self).collect()
+
+    def get_gc_stats(self):
+        stats = {}
+        if hasattr(gc, "get_stats"):  # only available in Python 3.4+
+            generations = gc.get_stats()
+            for i, generation in enumerate(generations):
+                stats["python.gc.gen{}.collections".format(i + 1)] = generation["collections"]
```

Comment:
Rather than including the generation number in the name, this is where labels are meant to be used. See `jvm.gc.count` for an example, under https://www.elastic.co/guide/en/apm/agent/java/current/metrics.html#metrics-jvm.

Reply:
@axw is there an example payload somewhere for this? This would mean I'd have to split up the metrics by label into 3 metric sets, right?

Reply:
Handily there is an example in the Java agent's javadoc:

https://github.com/elastic/apm-agent-java/blob/b8d4ee1e7de5420caab010cda5e50314e1522367/apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/MetricSet.java#L38-L42

> This would mean I'd have to split up the metrics by label into 3 metric sets, right?

Yes

Reply:
> >  This would mean I'd have to split up the metrics by label into 3 metric sets, right?

> Yes

But isn't that bad news for storage efficiency? IIRC, the more metrics per document, the more efficient it is to store. If we split up the generation numbers, we'll store 4 documents of GC data per collection interval (one per generation, plus an unlabled one).

Reply:
It's not _ideal_ for storage, but with a 30 second interval, this is in the order of a few megabytes of storage per day per agent.

Using labels gives the option of aggregating the values _and_ getting a breakdown per generation. In Metrics Explorer you would chart `python.gc.collections` and then optionally group by `labels.generation`.